File: src/IGBZ.Domain/Catalog/Product.cs
namespace IGBZ.Domain.Catalog;

/// <summary>
/// محصول یک تننت — با Variant ها (SKU/قیمت/موجودی/رزرو).
/// رزرو اتمیک موجودی در لایهٔ زیرساخت با <c>findOneAndUpdate</c> انجام می‌شود (سند بخش ۷.۳).
/// </summary>
public class Product : Entity, ITenantEntity
{
    public string TenantId { get; set; } = string.Empty;

    public string Name { get; set; } = string.Empty;

    /// <summary>نامک/اسلاگ برای URL — یکتا در سطح تننت.</summary>
    public string Slug { get; set; } = string.Empty;

    public string? Description { get; set; }

    public List<string> CategoryIds { get; set; } = new();

    public List<ProductVariant> Variants { get; set; } = new();

    public List<string> Images { get; set; } = new();

    public bool IsPublished { get; set; }

    /// <summary>حذف نرم — محصولات حذف‌شده در کاتالوگ نمایش داده نمی‌شوند.</summary>
    public bool Deleted { get; set; }

    public bool IsDigital { get; set; }

    /// <summary>برای محصولات دیجیتال — ارجاع به فایل دانلودی.</summary>
    public string? DownloadRef { get; set; }

    // ── سئو (تولید خودکار در فاز ۹) ──
    public string? MetaTitle { get; set; }
    public string? MetaDescription { get; set; }
    public string? MetaKeywords { get; set; }

    /// <summary>ترجمه‌های خودکار محصول (زبان‌های دیگر) — سند بخش ۱۴.</summary>
    public List<ProductTranslation> Translations { get; set; } = new();

    public ProductVariant? GetVariant(string sku) =>
        Variants.FirstOrDefault(v => string.Equals(v.Sku, sku, StringComparison.OrdinalIgnoreCase));

    public decimal GetLowestPriceToman() =>
        Variants.Count == 0 ? 0 : Variants.Min(v => v.PriceToman);

    public bool IsAvailable(string sku, int quantity)
    {
        var variant = GetVariant(sku);
        return variant != null && variant.AvailableQuantity >= quantity;
    }
}

/// <summary>ترجمهٔ خودکار نام/توضیحات محصول به یک زبان.</summary>
public class ProductTranslation
{
    public string LanguageCode { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public DateTime CreatedOnUtc { get; set; } = DateTime.UtcNow;
}

File: apps/admin_app/lib/models/product.dart
class ProductVariant {
  ProductVariant({
    required this.sku,
    this.attributes = const {},
    required this.priceToman,
    this.oldPriceToman,
    this.stockQuantity = 0,
  });

  final String sku;
  final Map<String, String> attributes;
  final double priceToman;
  final double? oldPriceToman;
  final int stockQuantity;

  factory ProductVariant.fromJson(Map<String, dynamic> json) => ProductVariant(
        sku: json['sku'] as String,
        attributes: (json['attributes'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, v.toString())),
        priceToman: (json['priceToman'] as num).toDouble(),
        oldPriceToman: (json['oldPriceToman'] as num?)?.toDouble(),
        stockQuantity: json['stockQuantity'] as int? ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'sku': sku,
        'attributes': attributes,
        'priceToman': priceToman,
        'oldPriceToman': oldPriceToman,
        'stockQuantity': stockQuantity,
      };
}

class Product {
  Product({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    this.images = const [],
    this.isPublished = false,
    this.variants = const [],
  });

  final String id;
  final String name;
  final String slug;
  final String? description;
  final List<String> images;
  final bool isPublished;
  final List<ProductVariant> variants;

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id'] as String,
        name: json['name'] as String,
        slug: json['slug'] as String,
        description: json['description'] as String?,
        images: (json['images'] as List<dynamic>? ?? []).cast<String>(),
        isPublished: json['isPublished'] as bool? ?? false,
        variants: (json['variants'] as List<dynamic>? ?? [])
            .map((e) => ProductVariant.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class ProductInput {
  ProductInput({
    required this.name,
    required this.slug,
    this.description,
    this.images = const [],
    this.isPublished = false,
    this.isDigital = false,
    this.variants = const [],
  });

  final String name;
  final String slug;
  final String? description;
  final List<String> images;
  final bool isPublished;
  final bool isDigital;
  final List<ProductVariant> variants;

  Map<String, dynamic> toJson() => {
        'name': name,
        'slug': slug,
        'description': description,
        'images': images,
        'isPublished': isPublished,
        'isDigital': isDigital,
        'variants': variants.map((v) => v.toJson()).toList(),
      };
}
